client/src/partials/receipts/receipt.sale.js

```diff
@@ -67,6 +68,25 @@ angular.module('bhima.controllers')
       }
     };
 
+    function updateCurrency (currency) {
+      var totals = model.total,
+        saleRecords = model.saleRecords[0];
+        
+      totals.localeCost = model.common.doConvert(saleRecords.cost, currency, saleRecords.invoice_date);
```

Comment:
The same logic as totals = model.total can be applied to almost every line of this function.  For example, you call `model.common.doConvert()` potentially 4 + length(model.saleRecords) times in updateCurrency()!  You clearly should cache this function in a local variable that can be called less expensively.